PMTUD

martinduke opened the pull request:
Rewrote the PMTUD section as discussed in Issue #64:
https://github.com/quicwg/base-drafts/issues/64

Because of concern about user-space implementations controlling kernel PMTU mechanisms, and interest in PLPMTUD as an alternative to tradition ICMP PMTUD, this is a thick web of different options, MUSTs and SHOULDs. In my truest heart I would like to sweep a lot of this away and do a MUST based on around the ICMP system, but though wordy this is more likely to please everyone.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,91 @@ The fields of a GOAWAY frame are as follows:
+risk by using the (at minimum) 8 bytes of transport header echoed in the ICMP message
+to validate the TCP sequence number as valid for the current connection. However, as
+QUIC operates over UDP, this echoed information may consist only of the IP and UDP
+headers, which hold no connection-specific context except the IP/port 4-tuple.
```
This isn't strictly correct.  There are 4 16-bit values (source port, destination port, length, checksum).  The problem is that not many implementations are willing to remember length and checksum for the packets that they sent.  Also, including length means that the packet contains very little entropy, meaning that it's relatively easy to guess from off-path.

martinduke replied:
That's a good point. How about "this echoed information may consist only of the IP and UDP headers, which usually has insufficient entropy to mitigate off-path attacks." ?

martinthomson replied:
Yeah, that works.

martinduke replied:
Changes reflected in patch-1.

martinduke commented:
I submitted patch-1, which is a major rewrite.

I realized that my PR was too long because it was extended digression on PMTU issues. So I shrunk it down to achieve a couple of things:
1) explicitly allow ICMP-based PMTUD
2) Codify the requirement to pad/DF-mark handshake packets.
3) List a few QUIC-specific considerations about ICMP PMTUD, given that many readers are coming from the TCP world where it is less fraught.

I believe this version either obviates or answers each comment on both #105 and #106.

martinduke commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,91 @@ The fields of a GOAWAY frame are as follows:
+in addition to, PMTUD. This avoids the off-path attack and misbehaving router problems
+while potentially introducing multiple packet losses and long timeouts. 
+
+Due to the drawbacks of PLPMTUD, QUIC endpoints MAY use ICMP-based PTMU discovery.
```
eliminated.

martinduke commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,91 @@ The fields of a GOAWAY frame are as follows:
+The last server-generated packet in the Transport Handshake SHOULD be padded to
+maximum-size with the DF bit set. Earlier packets may be smaller to save server
+resources until various handshake mechanisms have validated the client.
+
```
I made the edit.

martinduke commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,91 @@ The fields of a GOAWAY frame are as follows:
+## Packet MTU Determination
+
+QUIC endpoints SHOULD set the Don't Fragment (DF) bit in the IP header of selected
+QUIC datagrams. These packets MUST use PADDING frames, as necessary, to raise the
```
text eliminated

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,25 @@ The fields of a GOAWAY frame are as follows:
+Discovery ({{!RFC1191}}, {{!RFC1981}}) or Packetization Layer Path MTU
+Discovery ({{!RFC4821}}) for detecting the path's MTU, setting the maximum
+packet size appropriately, and storing the result of previous PMTU
+determinations. In the absence of these mechanisms,  the recommended default
```
nit: extra space after "mechanisms,"

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,25 @@ The fields of a GOAWAY frame are as follows:
+
+QUIC endpoints that implement any kind of MTU discovery SHOULD maintain a
+separate PMTU estimate for each IP address the peer is using in the
+connection. Endpoints SHOULD maintain an estimate for each combination of
```
Is this first sentence attempting to anticipate multipath or connection migration, where a connection may send to multiple IPs?

Is the second sentence implying that an endpoint should persist PMTU across connections?  If so, could you clarify slightly, possibly "Endpoints SHOULD maintain an estimate for each combination of local and remote IP addresses for use in future connections."?

martinduke replied:
Yes, this refers to multipath. It's not terribly clearly written. What it's trying to say that if you have N addresses and the peer has M addresses, you should keep at least M MTU estimates. It'd be even better, though, to keep M x N MTU estimates.

janaiyengar replied:
I think the second sentence subsumes the first one (a change in local IP is a change in the IP pair). I think the sentences can be combined: "QUIC endpoints that implement any kind of MTU discovery SHOULD maintain a separate PMTU estimate for each combination of local and remote IP addresses (as each pairing may have a different maximum MTU in the path)."

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,25 @@ The fields of a GOAWAY frame are as follows:
+local and remote IP addresses (as each pairing may have a different minimum
+MTU in the path).
+
+All handshake packets MUST include enough PADDING frames to bring the packet
```
You can only have one padding frame per packet, so maybe "All handshake packets MUST include a PADDING frame to bring the packet ..."?

janaiyengar replied:
+1, except that the transport draft defines a PADDING frame currently as simply the Frame Type 0, meaning that a PADDING frame is basically 1 byte long. I'll change this to be Type + 0-padding to the end of the packet.

janaiyengar replied:
I would switch the order of this paragraph with the previous one, to put requirements (MUST) before others.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,25 @@ The fields of a GOAWAY frame are as follows:
+MTU in the path).
+
+All handshake packets MUST include enough PADDING frames to bring the packet
+to the maximum size the endpoint is enforcing. Furthermore, all IPv4 handshake
```
On a practical note, some OS's(ie: MacOSX) don't allow userspace apps to set the DF bit.

martinduke replied:
Isn't this a problem for the requirement to pad the handshake packets? Without this bit, routers could just fragment and we haven't achieved anything.

ianswett replied:
I completely agree.  In practice, most people are on networks with largish(ie: >1400 byte) MTUs, but it's a highly suboptimal situation.  I'm fine with your wording, but it means we should ensure Apple plans to add support for the DF bit to MacOSX and iOS.  If I've just overlooked an existing interface, then all the better.

janaiyengar replied:
I would make this a SHOULD. We can't ensure that existing implementations will allow this -- I'm not sure what Win10 does, but I believe older versions of Windows don't support setting IP header bits.

martinduke commented:
I'm not sure why it's not showing up in this thread, but the latest version in patch-1 reflects the following:
- Added the clarification that packet number ambiguity is only a problem in IPv4
- Addressed Ian's comments, except...
- I kept the DF requirement for handshake packets, because I don't see how the existing Google implementation works without it.

ianswett commented:
This seems like a pragmatic change to the current docs.  We can iterate in the future, but it's much better than the status quo, so I'd like to see it merged before -01.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,25 @@ The fields of a GOAWAY frame are as follows:
+QUIC endpoints that implement any kind of MTU discovery SHOULD maintain a
+separate PMTU estimate for each IP address the peer is using in the
+connection. Endpoints SHOULD maintain an estimate for each combination of
+local and remote IP addresses (as each pairing may have a different minimum
```
nit: minimum --> maximum

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1270,27 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+insufficient entropy to mitigate off-path attacks.
+
+As a result, endpoints that implement PMTUD SHOULD take steps to mitigate this
+risk, which may include:
```
nit: "SHOULD take steps to mitigate this risk. For instance, an application may:"

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,23 @@ The fields of a GOAWAY frame are as follows:
-the maximum packet size appropriately.
+The maximum packet size for QUIC is the maximum size of the entire UDP payload,
+including the public header, encrypted payload, and any authentication fields.
+All QUIC packets SHOULD be sized to fit within the path's MTU to avoid IP
```
s/path's/estimated path/

martinthomson replied:
Separately, it might be wise to expand MTU on first use.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,23 @@ The fields of a GOAWAY frame are as follows:
+All QUIC packets SHOULD be sized to fit within the path's MTU to avoid IP
+fragmentation. To optimize bandwidth efficiency, endpoints MAY use Path MTU
+Discovery ({{!RFC1191}}, {{!RFC1981}}) or Packetization Layer Path MTU
+Discovery ({{!RFC4821}}) for detecting the path's MTU, setting the maximum
```
s/path's/path/ - the usual form drops the possessive.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,23 @@ The fields of a GOAWAY frame are as follows:
+Discovery ({{!RFC4821}}) for detecting the path's MTU, setting the maximum
+packet size appropriately, and storing the result of previous PMTU
+determinations. In the absence of these mechanisms, the recommended default
+maximum packet size is 1350 bytes for IPv6 and 1370 bytes for IPv4. 
```
I don't remember any discussion concluding that those numbers are OK.  Are we assuming 1398 bytes is available to IP?

I find it a little strange that this is accounting for the IP and UDP headers.  I usually see MTU discussed in terms of the entire IP packet.  I would phrase this differently: "Prior to attempting PMTUD, or if PMTUD is not attempted, an endpoint SHOULD assume a path MTU of 1398 octets.  Assuming that there are no IPv6 extension headers or IPv4 options in use, this results in a UDP payload length of 1350 octets for IPv6 and 1370 octets for IPv4."

Or is the intent to actually say that either extension headers are forbidden or that elements that intend to add extension headers have to be able to pass larger packets?

martinduke replied:
I agree with this fix and will make the change.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,23 @@ The fields of a GOAWAY frame are as follows:
+determinations. In the absence of these mechanisms, the recommended default
+maximum packet size is 1350 bytes for IPv6 and 1370 bytes for IPv4. 
+
+All handshake packets MUST include a PADDING frame if necessary to bring the
```
Is this ALL packets?  If the server needs to send multiple handshake packets (that happens often because certificates get bloated), do they really need to pad the 10th packet they send?  Surely the goal has been achieved after sending the first packet at the maximum size.  (The client should only ever send one handshake packet.)

martinduke replied:
I followed a suggestion in this thread to make it all packets. It's also hard to specify individual packets in the handshake because that's not part of this document. But I'll make it "at least" one packet in the handshake.

ianswett replied:
One word of warning.  If you pad the first CHLO, but not the retransmission of that, bad things happen.  So I'd rather be safe and say all packets because it's simple and hard to get wrong.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,23 @@ The fields of a GOAWAY frame are as follows:
+maximum packet size is 1350 bytes for IPv6 and 1370 bytes for IPv4. 
+
+All handshake packets MUST include a PADDING frame if necessary to bring the
+packet to the maximum size the endpoint is enforcing. Furthermore, all IPv4
```
The phrasing here is odd.  Which endpoint is enforcing a size?  Endpoints don't really enforce limits, it's something on the path that is doing any enforcement.  I would say that to the estimated maximum packet size for the path.

Given that we don't have an estimate for the path at this point, why not use the defaults?  Or are we keeping the option open to remember path MTU from a previous use of that path?  FWIW, I would rather we *explicitly* enabled that as a feature if we decide that it's OK.  I've had bad experiences with smuggling features into protocols, even seemingly innocuous ones.

martinduke replied:
The PMTU RFCs explictly discussing storing PMTU results from previous connections. I don't think it's "smuggling" to not enumerate the components of those RFCs, nor do I think the language above implies a choice of either active PMTUD or just using the defaults.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,23 @@ The fields of a GOAWAY frame are as follows:
+
+QUIC endpoints that implement any kind of MTU discovery SHOULD maintain an
+estimate for each combination of local and remote IP addresses (as each pairing
+may have a different maximum MTU in the path).
```
s/may/could

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1268,27 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+## Special Considerations for Path MTU Discovery 
+
+Traditional ICMP-based path MTU discovery in IPv4 ({{!RFC1191}} is potentially
+vulnerable to off-path attacks that succesfully guess the IP/port 4-tuple and
```
s/succesfully/successfully

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1268,27 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+risk by using the (at minimum) 8 bytes of transport header echoed in the ICMP
+message to validate the TCP sequence number as valid for the current
+connection. However, as QUIC operates over UDP, in IPv4 the echoed information
+may consist only of the IP and UDP headers, which usually has insufficient
```
s/may consist/consists ?  I don't know if there are cases where you get more or less data, are there?

martinduke replied:
Routers have the option of echoing more than 28 bytes of IP packet, and I've seen pcaps where this is the case.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1268,27 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+entropy to mitigate off-path attacks.
+
+As a result, endpoints that implement PMTUD in IPv4 SHOULD take steps to mitigate
+this risk. For instance, an application may:
```
Here you use "may" when talking about a SHOULD, which will confuse some people.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1268,27 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+
+* Set the IPv4 Don't Fragment (DF) bit on a small number of packets per RTT, so
+that most invalid ICMP messages arrive when there are no DF packets
+outstanding.
```
Maybe just  "a small proportion of packets".  If you DF on every RTT as this implies, then you will pretty much always have a DF packet outstanding.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1268,27 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+outstanding.
+
+* Store additional information from the IP or UDP headers (for example, the IP
+ID or UDP checksum) to further authenticate incoming Packet Too Big messages.
```
Maybe just "store the first IP and UDP headers from DF packets" to further authenticate [...]"

ianswett commented:
One comment on the importance of padding more than 1 packet in the handshake.

martinduke commented:
I believe this PR is close to ready. The main open issue, I believe, is what the specification says about padding handshake packets.

1) All handshake packets to be padded to maximum size. 
2) Pad at least one packet per handshake (in each direction) _that is successfully acknowledged_
3) More specifically, the first client packet (CHLO) and retransmissions, and last server packet [SHLO]  and retransmissions. In many cases this will be the same as #2, but is a bit more prescriptive.

martinduke commented:
I'm having trouble finding this comment on the GitHub page, but I believe
that the padding requirement is the only outstanding issue, and laid out
the options in another comment.

On Sun, Jan 15, 2017 at 8:09 PM, Martin Thomson <notifications@github.com>
wrote:

> *@martinthomson* commented on this pull request.
> ------------------------------
>
> In draft-ietf-quic-transport.md
> <https://github.com/quicwg/base-drafts/pull/106>:
>
> > -the resulting UDP datagram.  All QUIC packets SHOULD be sized to fit within the
> -path's MTU to avoid IP fragmentation.  The recommended default maximum packet
> -size is 1350 bytes for IPv6 and 1370 bytes for IPv4.  To optimize better,
> -endpoints MAY use PLPMTUD {{!RFC4821}} for detecting the path's MTU and setting
> -the maximum packet size appropriately.
> +The maximum packet size for QUIC is the maximum size of the entire UDP payload,
> +including the public header, encrypted payload, and any authentication fields.
> +All QUIC packets SHOULD be sized to fit within the path's MTU to avoid IP
> +fragmentation. To optimize bandwidth efficiency, endpoints MAY use Path MTU
> +Discovery ({{!RFC1191}}, {{!RFC1981}}) or Packetization Layer Path MTU
> +Discovery ({{!RFC4821}}) for detecting the path's MTU, setting the maximum
> +packet size appropriately, and storing the result of previous PMTU
> +determinations. In the absence of these mechanisms, the recommended default
> +maximum packet size is 1350 bytes for IPv6 and 1370 bytes for IPv4.
> +
> +All handshake packets MUST include a PADDING frame if necessary to bring the
>
> I'd consider retransmission of an initial packet to be equivalent to
> starting over. I'm concerned that we're going to affect the multi-packet
> server flight unnecessary, that's all.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/quicwg/base-drafts/pull/106>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXRMEcQ84m15PlxO-GHxS0a863LToxrSks5rSu1ggaJpZM4LXJpx>
> .
>

ianswett commented:
(1) seems simpler, but (3) seems both correct and reasonable as well.

martinthomson commented:
Can we split the discussion about padding handshake packets from this change?  That would let us merge this and allow discussion to continue.

martinduke commented:
OK, the handshake padding issue is gone from this PR. If there are no further objections, I believe the editors can merge this into the base draft.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,45 @@ The fields of a GOAWAY frame are as follows:
+
+In the absence of these mechanisms, QUIC endpoints SHOULD use a default PMTU of
+1280 octets. Assuming the minimum IP header size, this results in a UDP payload
+length of 1232 octets for IPv6 and 1252 octets for IPv4.
```
Given that 1280 is the minimum, this really isn't saying that there is a default PMTU, it's more saying that QUIC implementations can assume that 1280 is OK.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,45 @@ The fields of a GOAWAY frame are as follows:
+below 1280 octets. This is equal to the IPv6 minimum. Most modern IPv4 networks
+support at least this packet size. A large packet size allows the first client
+handshake packet to fit in a single datagram, which in turn enables stateless
+rejects by the server.
```
Nothing in the protocol will stop a QUIC endpoint from using a path with a smaller MTU, especially if they can't set DF.  So I would frame this differently:

> QUIC depends on the network path supporting a MTU of at least 1280 octets.  This is the IPv6 minimum and therefore also supported by most modern IPv4 networks.  An endpoint MUST NOT reduce their MTU below this number, even if it receives signals that indicate a smaller limit might exist.

Then I would follow with the ClientHello paragraph:

> A client MUST construct an initial packet that is at least 1280 octets in size, including IP and UDP headers.  It is RECOMMENDED that a packet be padded to exactly 1280 octets unless the client has a reasonable assurance that the PMTU is larger.  A server MUST reject initial packets that are less than 1280 octets in size (caveat warning: see below).  Sending a packet of this size ensures that the network path supports an MTU of this size and helps mitigate amplification attacks caused by server responses toward an unverified client address.

Then I would include the text on PMTU changes, which isn't as important.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1198,12 +1198,45 @@ The fields of a GOAWAY frame are as follows:
+PMTU through out-of-band information.
+
+Servers MUST reject the first plaintext packet received from a client if it its
+total size is less than 1280 octets, to mitigate amplification attacks.
```
Policing this will be difficult, if not impossible, if there are extension headers.  Userspace implementations might not even learn about them.  From what I can tell, there is no limit to the size of extension headers in IPv6, so we can't really require that this enforcement happen.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1290,31 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+this risk. For instance, an application could:
+
+* Set the IPv4 Don't Fragment (DF) bit on a small proportion of packets, so that
+most invalid ICMP messages arrive when there are no DF packets outstanding.
```
This is missing the (obvious) and can therefore be easily identified as spurious.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1257,6 +1290,31 @@ frames), the receiving peer MAY send an ACK frame after a reasonable number
+
+* Store additional information from the IP or UDP headers from DF packets (for
+example, the IP ID or UDP checksum) to further authenticate incoming Packet Too
+Big messages.
```
RFC 1191 uses "Datagram Too Big" rather than "Packet Too Big".

martinduke commented:
> Policing this will be difficult, if not impossible, if there are extension headers. Userspace implementations might not even learn about them. From what I can tell, there is no limit to the size of extension headers in IPv6, so we can't really require that this enforcement happen.

While I agree that the IP header length is somewhat unpredictable, a user space implementation that doesn't know what's going on will presumably assume the minimum IP header length, and the packet will end up too big, not big enough.

All of the 1st-packet restrictions are meaningless if the implementation is allowed to assume an IP header length > minimum without knowing that that is the case.

Other fixes are in the following commit, though I did not take Martin's suggestions verbatim in every case.

martinthomson commented:
Yes, I will open an issue for the one thing I raised.
